### ISSUE TITLE
Can not move to the previous error with flycheck-previous-error.

### ISSUE THREAD

**Issue by templateK** [flycheck/flycheck#1779]
## Bug description

  We can not move to the previous error with `flycheck-previous-error`
  from the error which has a mismatch between warning underline position
  and actual compiler parse error position.

## Steps to reproduce

1. Open file below code with `haskell-mode` and `flycheck-mode` enabled

    ```haskell
        module Main where

        main :: IO ()
        main = do
          let x = a
          let y = 000b000
          pure ()
    ```
2. Try to use `flycheck-previous-error` on the second error position. 

3. `flycheck-previous-error` does nothing.

## Expected behavior

  The cursor should be at the first error.

## System configuration
```
Syntax checkers for buffer ghc_err_report.hs in haskell-tng-mode:

First checker to run:

  lsp-ui (explicitly selected)
    - may enable: yes
    - may run:    t

Flycheck Mode is enabled.  Use M-x universal-argument C-c ! x to
enable disabled checkers.

--------------------

Flycheck version: 32snapshot (package: 20200516.1719)
Emacs version:    27.0.90
System:           x86_64-apple-darwin18.6.0
Window system:    mac
```

Emacs configuration:
- [x] Plain Emacs / Custom configuration

**Comment by cpitclaudel** [comment]
Thanks.  Can you give a recipe from emacs -Q?

**Comment by templateK** [comment]
If you already have a ghc compiler on your system, skip 1 and 2.

1. Download proper ghc binary package for your OS from [here](https://www.haskell.org/ghc/download_ghc_8_10_1.html) and extract it somewhere.


2. Add ghc binary path to `$PATH` and make sure you have ghc compiler on your path. 
  this can be confirmed by `ghc --version` output should be`The Glorious Glasgow Haskell Compilation System, version x.xx.x`

3. Evalute the following code after `emacs -Q`

```elisp
(require 'package)

(defun prep-pkgs ()
  (setq package-list '(haskell-mode flycheck))
  (setq package-archives '(("melpa" . "https://melpa.org/packages/")))
  (package-initialize)
  (unless package-archive-contents  (package-refresh-contents))
  (dolist (package package-list)
    (unless (package-installed-p package) (package-install package))))

(prep-pkgs)

(global-set-key (kbd "<f1>") 'flycheck-mode)
(global-set-key (kbd "<f2>") (lambda () (interactive) (flycheck-previous-error)))
(global-set-key (kbd "<f3>") (lambda () (interactive) (flycheck-next-error))))

(find-file "foo.hs")
```
after that, paste the [sample](https://gist.github.com/templateK/f8924b5bf72cbd0fa81b6c03c237f74f)  and enable `flycheck-mode` manually by pressing `<f1>`

**Comment by cpitclaudel** [comment]
Thanks, this recipe is perfect.  Will look into it.

**Comment by templateK** [comment]
I forgot one step between 1 and 2.  ghc binary package installation process.

1.5 After downloading the tarball and extracting it, you have to do the following
```
cd <tarball extracted directory>
./configure --prefix=/absolute/path/to/install
```
and ghc binary will be at `/absolute/path/to/install/bin/ghc`

**Comment by cpitclaudel** [comment]
> I forgot one step between 1 and 2. ghc binary package installation process.

Yup, I figured that one out, no sweat.  The recipe was top notch.

I think https://github.com/flycheck/flycheck/tree/cpitclaudel_1519-fix-previous-error should fix the issue, but I want to update the test suite to catch this before merging.

**Comment by cpitclaudel** [comment]
In particular, even with this patch, navigating over nested errors is still broken.

**Comment by templateK** [comment]
I don't understand what "nested errors" means exactly.  Can you provide a example? 

And about merging, no need to hurry. I will  use your commit by patching the flycheck manually in the mean time.

**Comment by cpitclaudel** [comment]
> I don't understand what "nested errors" means exactly. Can you provide a example?

In OCaml, the following snippet has 2 errors:

```
let x = true && (1 + false)
```

This has two errors, one nested into the other:

```
‘(1 + false)’: This expression has type int but an expression was expected of type bool
‘false’: This expression has type bool but an expression was expected of type int
```

**Comment by cpitclaudel** [comment]
I've pushed a slightly different fix; see #1780.  I also opened a separate ticket about error navigation https://github.com/flycheck/flycheck/issues/1781

**Comment by templateK** [comment]
> In OCaml, the following snippet has 2 errors:
> 
> ```
> let x = true && (1 + false)
> ```
> 
> This has two errors, one nested into the other:
> 
> ```
> ‘(1 + false)’: This expression has type int but an expression was expected of type bool
> ‘false’: This expression has type bool but an expression was expected of type int
> ```

Thanks for the example.

FYI, I can replicate sort of similar situation in haskell too.

```haskell
  let x = "foo" ++ ((1 + False) :: Int)
                    ~~
```
error message is following.
  1.  ``` •  Couldn't match expected type ‘[Char]’ with actual type ‘Int’```
  2. ``` •  Couldn't match expected type ‘Int’ with actual type ‘Bool’```

and we can not move from 2 to 1.

I personally don't mind #1780 isn't working for the above case, but it certainly resolves all my 
original case. so, I think we shoud keep this issue open until  this issue can be fully resolved.

**Comment by cpitclaudel** [comment]
I opened #1781 to track that case; could you add your example there too? :) This way we can keep this to previous-error with highlighting-mode on symbols.

Thanks again for reporting this, btw.  I don't use these navigation commands often and I hadn't noticed the issue(s).

**Comment by templateK** [comment]
I've posted https://github.com/flycheck/flycheck/issues/1781#issuecomment-631798504 .
We can continue discussion about nested errors navigation from #1781 if something better comes up, and it's fair to say that closing this issue is your choice. 

Thank you for the awesome package and the hard work.
